src/ensembl/src/content/home/homePageSelectors.ts

```diff
@@ -0,0 +1,91 @@
+import * as urlFor from 'src/shared/helpers/urlHelper';
+import { getChrLocationStr } from 'src/content/app/browser/browserHelper';
+
+import {
+  getBrowserActiveEnsObjectIds,
+  getAllChrLocations
+} from 'src/content/app/browser/browserSelectors';
+import {
+  getEnsObjectById,
+  getEnsObjectLoadingStatus
+} from 'src/ens-object/ensObjectSelectors';
+import { getCommittedSpecies } from 'src/content/app/species-selector/state/speciesSelectorSelectors';
+import { LoadingState } from 'src/shared/types/loading-state';
+
+import { RootState } from 'src/store';
+
+export type PreviouslyViewedGenomeBrowserObject = {
+  speciesName: string;
+  assemblyName: string;
+  link: string;
+};
+
+export type PreviouslyViewedGenomeBrowserObjects = {
+  areLoading: boolean;
+  objects: PreviouslyViewedGenomeBrowserObject[];
+};
+
+export const getPreviouslyViewedGenomeBrowserObjects = (
+  state: RootState
+): PreviouslyViewedGenomeBrowserObjects => {
+  const ensObjectIdsMap = getBrowserActiveEnsObjectIds(state);
+  const chrLocations = getAllChrLocations(state);
+  const committedSpecies = getCommittedSpecies(state);
+
+  const ensObjectMap: {
+    [genomeId: string]: {
+      speciesName: string;
+      assemblyName: string;
+      ensObjectId: string;
+    };
+  } = Object.keys(ensObjectIdsMap).reduce((result, genomeId) => {
+    const ensObjectId = ensObjectIdsMap[genomeId];
+    const ensObject = getEnsObjectById(state, ensObjectId);
+    const species = committedSpecies.find(
+      (species) => species.genome_id === genomeId
+    );
+
+    if (!ensObject || !species) {
+      return result;
+    } else {
+      const speciesName = species.common_name || species.scientific_name;
+      const assemblyName = species.assembly_name;
+      const ensObjectId = ensObject.object_id;
+      return {
+        ...result,
+        [genomeId]: {
+          speciesName,
+          assemblyName,
+          ensObjectId
+        }
+      };
+    }
+  }, {});
+
+  const areLoading: boolean = Object.keys(ensObjectIdsMap).reduce(
```

Comment:
Is using `reduce` a good option here? If the first object is still loading, we don't really have to process the rest.

Reply:
What alternative approaches do you have in mind? We don't have a `reduceWhile` (as Ramda has); and I would much rather use a reduce than a for-loop with external flags indicating whether or not to continue iteration. Execution of this iteration is going to be very fast (there are only a few committed species a user can reasonably have, and for each species it's just a lookup of data in a hash table).

Reply:
We can instead use `every`?

Reply:
I agree that these blocks run very fast on our machines but there will be some slower systems that could work slightly better by avoiding the few extra iterations. 
I just did a site speed test and it says our website takes 3.6 seconds to load on 4G, which is already considered to be slow by google:
https://www.thinkwithgoogle.com/intl/en-gb/feature/testmysite/

I am worried that this value will only go higher as we keep adding features without considering these minor things.

Reply:
> I just did a site speed test and it says our website takes 3.6 seconds to load on 4G

Loading time (on any G) depends on the amount of bytes that have to be downloaded. It does not depend on whether you run a longer or a shorter cycle :-)

In order to check the impact of iterating over a list of items, you would need to profile a specific piece of code. I am sure there would be no significant difference even if we had to loop over a thousand items; but in this case the number of items to iterate over will at most be a couple of dozens.

Following @imran-ebi 's suggestion, I replaced `reduce` with `some`.